README.md
# ⚠️ WARNING ⚠️

Telegraf now has an official Internet Speed Monitor plugin. It doesn't record as much data as this container, but it is officially supported, if that matters to you.

https://github.com/influxdata/telegraf/tree/master/plugins/inputs/internet_speed

# docker-speedtest-influxdbv2

[![CI/CD](https://github.com/gibby/docker-speedtest-influxdbv2/actions/workflows/main.yml/badge.svg)](https://github.com/gibby/docker-speedtest-influxdbv2/actions/workflows/main.yml)
[![Docker Image Size (latest semver)](https://img.shields.io/docker/image-size/gibby/docker-speedtest-influxdbv2)](https://hub.docker.com/r/gibby/docker-speedtest-influxdbv2)

Runs Ookla's [Speedtest CLI](https://www.speedtest.net/apps/cli) program in Docker, sends the results to InfluxDB
  - Source code: [GitHub](https://github.com/gibby/docker-speedtest-influxdbv2)
  - Dockerhub container: [Docker Hub](https://hub.docker.com/r/gibby/docker-speedtest-influxdbv2)
  - Image base: [Python (slim Buster)](https://hub.docker.com/_/python)
  - Init system: N/A
  - Application: [Speedtest CLI](https://www.speedtest.net/apps/cli)
  - Architecture: `linux/amd64`


## Explanation

  - This runs Ooka's Speedtest CLI program on an interval, then writes the data to an InfluxDB database (you can later graph this data with Grafana or Chronograf)
  - This does **NOT** use the open-source [speedtest-cli](https://github.com/sivel/speedtest-cli). That program uses the Speedtest.net HTTP API. This program uses Ookla's official CLI application.
  - ⚠️ Ookla's speedtest application is closed-source (the binary applications are [here](https://www.speedtest.net/apps/cli)) and Ookla's reasoning for this decision is [here](https://www.reddit.com/r/HomeNetworking/comments/dpalqu/speedtestnet_just_launched_an_official_c_cli/f5tm9up/) ⚠️
  - ⚠️ Ookla's speedtest application reports all data back to Ookla ⚠️
  - ⚠️ This application uses Ookla's recommendation to install by piping curl to bash  ⚠️
  - The default output unit of measurement is **bytes-per-second**. You will most likely want to convert to megabits-per-second by dividing your output by 125000.

```
For example, if your download speed is 11702913 bytes-per-second:
11702913 / 125000 = 93.623304 megabits-per-second
```

## Requirements

  - This only works with InfluxDB v1.8 and v2, because I'm using [this](https://github.com/influxdata/influxdb-client-python) client library.
  - You must already have an InfluxDB database created, along with a user that has `WRITE` and `READ` permissions on that database.
  - This Docker container needs to be able to reach that InfluxDB instance by hostname, IP address, or Docker service name (I run this container on the same Docker network as my InfluxDB instance).
  - ⚠️ Depending on how often you run this, you may need to monitor your internet connection's usage. If you have a data cap, you could exceed it. The standard speedtest uses about 750MB of data per run. See below for an example. ⚠️

```
CONTAINER: NET I/O
speedtest: 225MB / 495MB
```

## Docker image information

### Docker image tags
  - `latest`: Latest version
  - `X.X.X`: [Semantic version](https://semver.org/) (use if you want to stick on a specific version)

### Environment variables
| Variable         | Required?                  | Definition                                     | Example                                     | Comments                                                                                            |
|------------------|----------------------------|------------------------------------------------|---------------------------------------------|-----------------------------------------------------------------------------------------------------|
| INFLUXDB_SCHEME  | No (default: http)         | Connect to InfluxDB using http or https        | 'https'                                     | Useful if InfluxDB is behind a reverse proxy and you need to use https                              |
| INFLUXDB_HOST    | No (default: localhost)    | Server hosting the InfluxDB                    | 'localhost' or your Docker service name     |                                                                                                     |
| INFLUXDB_PORT    | No (default: 8086)         | InfluxDB port                                  | 8086                                        |                                                                                                     |
| INFLUXDB_USER    | Yes (only for v1)          | Database username                              | test1                                       | Needs to have the correct permissions                                                               |
| INFLUXDB_PASS    | Yes (only for v1)          | Database password                              | test1234                                    |                                                                                                     |
| INFLUXDB_TOKEN   | Yes (only for v2)          | Token to connect to bucket                     | asdfghjkl                                   | Needs to have the correct permissions. Setting this assumes we're talking to an InfluxDBv2 instance |
| INFLUXDB_ORG     | Yes (only for v2)          | Organization                                   | my_test_org                                 |                                                                                                     |
| INFLUXDB_DB      | Yes                        | Database name                                  | SpeedtestStats                              | Must already be created. In InfluxDBv2, this is the "bucket".                                       |
| SLEEPY_TIME      | No (default: 3600)         | Seconds to sleep between runs                  | 3600                                        | The loop takes about 15-30 seconds to run, so I wouldn't set this value any lower than 60 (1min)    |
| SLEEPY_TIME_RANDOM | No (default: false) | Randomizes the sleepy time +/- 20% | true | Useful when running as a daemon set
| SPEEDTEST_HOST   | No (default: container ID) | Hostname of service where Speedtest is running | server04                                    | Useful if you're running Speedtest on multiple servers                                              |
| SPEEDTEST_SERVER | No (default: random)       | ID number of Speedtest server                  | 41817                                       | See a list of servers and IDs [here](https://c.speedtest.net/speedtest-servers-static.php)          |

### Ports
N/A

### Volumes
N/A

### Example usage
Below is an example docker-compose.yml file for connecting to InfluxDB v2.
```
version: '3'
services:
  speedtest2:
    container_name: docker-speedtest-influxdbv2
    restart: unless-stopped
    environment:
      - INFLUXDB_SCHEME=http
      - INFLUXDB_HOST=influxdbv2
      - INFLUXDB_PORT=8086
      - INFLUXDB_TOKEN=asdfghjkl
      - INFLUXDB_ORG=my_test_org
      - INFLUXDB_DB=SpeedtestStats
      - SLEEPY_TIME=3600
      - SPEEDTEST_HOST=server04
      - SPEEDTEST_SERVER=41817
    networks:
      - speedtest2
    image: gibby/docker-speedtest-influxdbv2:latest

networks:
  speedtest2:
```

Below is an example docker-compose.yml file for connecting to InfluxDB v1.8.
```
version: '3'
services:
  speedtest1:
    container_name: docker-speedtest-influxdbv1
    restart: unless-stopped
    environment:
      - INFLUXDB_SCHEME=http
      - INFLUXDB_HOST=influxdbv1
      - INFLUXDB_PORT=8086
      - INFLUXDB_USER=test1
      - INFLUXDB_PASS=test1234
      - INFLUXDB_DB=SpeedtestStats
      - SLEEPY_TIME=3600
      - SPEEDTEST_HOST=server04
      - SPEEDTEST_SERVER=41817
    networks:
      - speedtest1
    image: gibby/docker-speedtest-influxdbv2:latest

networks:
  speedtest1:
```

Below is an example of running locally (used to edit/test/debug).
```
# Build the Dockerfile
docker compose -f docker-compose-dev.yml up -d

# View logs
docker compose -f docker-compose-dev.yml logs -f

# Destroy when done
docker compose -f docker-compose-dev.yml down
```

## TODO
- [ ] Learn Python
- [x] ~~Run the processes inside the container as a non-root user~~
- [ ] Add a [healthcheck](https://docs.docker.com/engine/reference/builder/#healthcheck)
- [x] ~~Move the database connection check to a function~~
- [x] ~~Add logic to check if variables are set~~
- [x] ~~Add defaults for HOST and PORT~~
- [ ] Update CI/CD with tests
- [x] ~~Add warning about bandwidth~~
- [ ] Implement an authentication check when [this PR](https://github.com/influxdata/influxdb-client-python/pull/269) is merged


CHANGELOG.md
# Changelog

## [1.3.1](https://github.com/Gibby/docker-speedtest-influxdbv2/compare/v1.3.0...v1.3.1) (2023-08-16)


### Bug Fixes

* My first release ([#1](https://github.com/Gibby/docker-speedtest-influxdbv2/issues/1)) ([bcb465a](https://github.com/Gibby/docker-speedtest-influxdbv2/commit/bcb465a93825f5b6a62da000ed3393d5966506bd))


app/speedtest.py
#!/usr/bin/env python3

import datetime
import json
import os
import subprocess  # nosec B404
import time
import socket
import sys
from influxdb_client import InfluxDBClient
import random

# Variables
influxdb_scheme = os.getenv("INFLUXDB_SCHEME", "http")
influxdb_host = os.getenv("INFLUXDB_HOST", "localhost")
influxdb_port = int(os.getenv("INFLUXDB_PORT", 8086))
influxdb_user = os.getenv("INFLUXDB_USER")
influxdb_pass = os.getenv("INFLUXDB_PASS")
influxdb_token = os.getenv("INFLUXDB_TOKEN")
influxdb_org = os.getenv("INFLUXDB_ORG", "-")
influxdb_db = os.getenv("INFLUXDB_DB")
sleepy_time = int(os.getenv("SLEEPY_TIME", 3600))
sleepy_time_random = bool(os.getenv("SLEEPY_TIME_RANDOM", False))
sleepy_time_do = int("0")
start_time = datetime.datetime.utcnow().isoformat()
default_hostname = socket.gethostname()
hostname = os.getenv("SPEEDTEST_HOST", default_hostname)
speedtest_server = os.getenv("SPEEDTEST_SERVER")
debug_mode = bool(os.getenv("DEBUG_MODE", False))


def db_check():

    if debug_mode is True:
        print("STATE: Debug mode is enabled, skipping the database check!")
    else:
        print("STATE: Running database check")
        client_health = client.ping()

        if client_health is True:
            print("STATE: Connection", client_health)
        elif client_health is False:
            print("ERROR: Connection", client_health, " - Check scheme, host, port, user, pass, token, org, etc...")
            sys.exit(1)
        else:
            print("ERROR: Something else went wrong")
            sys.exit(1)


def speedtest():
    db_check()

    current_time = datetime.datetime.utcnow().isoformat()
    print("STATE: Loop running at", current_time)

    # Run Speedtest
    # If the user specified a speedtest_server ID number, run a different command vs if they didn't specify an ID
    if speedtest_server:
        print("STATE: User specified speedtest server:", speedtest_server)
        speedtest_server_arg = "--server-id="+speedtest_server
        print("STATE: Speedtest running")
        my_speed = subprocess.run(['/usr/bin/speedtest', '--accept-license', '--accept-gdpr', '--format=json', speedtest_server_arg], stdout=subprocess.PIPE, shell=False, text=True, check=True)  # nosec B603
    else:
        print("STATE: User did not specify speedtest server, using a random server")
        print("STATE: Speedtest running")
        my_speed = subprocess.run(['/usr/bin/speedtest', '--accept-license', '--accept-gdpr', '--format=json'], stdout=subprocess.PIPE, shell=False, text=True, check=True)  # nosec B603

    # Convert the string into JSON, only getting the stdout and stripping the first/last characters
    my_json = json.loads(my_speed.stdout.strip())

    # Get the values from JSON and log them to the Docker logs
    # Basic values
    speed_down = my_json["download"]["bandwidth"]
    speed_up = my_json["upload"]["bandwidth"]
    ping_latency = my_json["ping"]["latency"]
    ping_jitter = my_json["ping"]["jitter"]
    result_url = my_json["result"]["url"]
    # Advanced values
    speedtest_server_id = my_json["server"]["id"]
    speedtest_server_name = my_json["server"]["name"]
    speedtest_server_location = my_json["server"]["location"]
    speedtest_server_country = my_json["server"]["country"]
    speedtest_server_host = my_json["server"]["host"]

    # Print results to Docker logs
    print("STATE: RESULTS ARE SAVED IN BYTES-PER-SECOND NOT MEGABITS-PER-SECOND")
    print("STATE: Your download     ", speed_down, "B/s")
    print("STATE: Your upload       ", speed_up, "B/s")
    print("STATE: Your ping latency ", ping_latency, "ms")
    print("STATE: Your ping jitter  ", ping_jitter, "ms")
    print("STATE: Your server info  ", speedtest_server_id, speedtest_server_name, speedtest_server_location, speedtest_server_country, speedtest_server_host)
    print("STATE: Your URL is       ", result_url)

    # This is ugly, but trying to get output in line protocol format (UNIX time is appended automatically)
    # https://docs.influxdata.com/influxdb/v2.0/reference/syntax/line-protocol/
    p = "speedtest," + "service=speedtest.net," + "host=" + str(hostname) + " download=" + str(speed_down) + ",upload=" + str(speed_up) + ",ping_latency=" + str(ping_latency) + ",ping_jitter=" + str(ping_jitter) + ",speedtest_server_id=" + str(speedtest_server_id) + ",speedtest_server_name=" + "\"" + str(speedtest_server_name) + "\"" + ",speedtest_server_location=" + "\"" + str(speedtest_server_location) + "\"" + ",speedtest_server_country=" + "\"" + str(speedtest_server_country) + "\"" + ",speedtest_server_host=" + "\"" + str(speedtest_server_host) + "\"" + ",result_url=" + "\"" + str(result_url) + "\""
    # For troubleshooting the raw line protocol
    # print(p)

    if debug_mode is True:
        print("STATE: Debug mode is enabled, skipping writing to the database!")
        print(p)
    else:
        try:
            print("STATE: Writing to database")
            write_api = client.write_api()
            write_api.write(bucket=influxdb_db, record=p)
            write_api.__del__()
        except Exception as err:
            print("ERROR: Error writing to database")
            print(err)
    if sleepy_time_random is True:
        sleepy_time_start = sleepy_time * 0.2
        sleepy_time_end = sleepy_time * 1.2
        sleepy_time_do = random.randint(sleepy_time_start, sleepy_time_end)  # nosec B311
    else:
        sleepy_time_do = sleepy_time
    print("STATE: Sleeping for", sleepy_time_do, "seconds")
    time.sleep(sleepy_time_do)


# Some logging
print("#####\nScript starting!\n#####")
print("STATE: Starting at", start_time)
print("STATE: Sleep time between runs set to", sleepy_time_do, "seconds")

# Check if variables are set
print("STATE: Checking environment variables...")

if debug_mode is True:
    print("STATE: Debug mode is enabled, not checking for any variables!")
else:
    if 'INFLUXDB_DB' in os.environ:
        print("STATE: INFLUXDB_DB is set")
        pass
    else:
        print("ERROR: INFLUXDB_DB is not set")
        sys.exit(1)

    if 'INFLUXDB_TOKEN' in os.environ:
        print("STATE: INFLUXDB_TOKEN is set, so we must be talking to an InfluxDBv2 instance")
        pass
        # If token is set, then we are talking to an InfluxDBv2 instance, so INFLUXDB_ORG must also be set
        if 'INFLUXDB_ORG' in os.environ:
            print("STATE: INFLUXDB_ORG is set")
            pass
        else:
            print("ERROR: INFLUXDB_TOKEN is set, but INFLUXDB_ORG is not set")
            sys.exit(1)
    else:
        print("STATE: INFLUXDB_TOKEN is not set, so we must be talking to an InfluxDBv1 instance")
        # If token is not set, then we are talking an InfluxDBv1 instance, so INFLUXDB_USER and INFLUXDB_PASS must also be set
        if 'INFLUXDB_USER' in os.environ:
            print("STATE: INFLUXDB_USER is set")
            pass
        else:
            print("ERROR: INFLUXDB_USER is not set")
            sys.exit(1)

        if 'INFLUXDB_PASS' in os.environ:
            print("STATE: INFLUXDB_PASS is set")
            pass
        else:
            print("ERROR: INFLUXDB_PASS is not set")
            sys.exit(1)
        # If token is not set, influxdb_token must be a concatenation of influxdb_user:influxdb_pass when talking to an InfluxDBv1 instance
        # https://docs.influxdata.com/influxdb/v1.8/tools/api/#apiv2query-http-endpoint
        influxdb_token = f'{influxdb_user}:{influxdb_pass}'

    # Instantiate the connection
    connection_string = influxdb_scheme + "://" + influxdb_host + ":" + str(influxdb_port)
    print("STATE: Database URL is... " + connection_string)
    print("STATE: Connecting to InfluxDB...")
    client = InfluxDBClient(url=connection_string, token=influxdb_token, org=influxdb_org)

while True:
    speedtest()
